### ISSUE TITLE
[ISSUE] Missing XMP Metadata

### ISSUE THREAD

**Issue by rosx27** [pikepdf/pikepdf#537]
i'm creating a document which should include the metadata, below is the code snippet I have so far:
```
from fastapi import FastAPI, File, UploadFile, Response, Form, HTTPException
from fastapi.responses import FileResponse, RedirectResponse, HTMLResponse
from fastapi.staticfiles import StaticFiles
from tempfile import NamedTemporaryFile
from io import BytesIO
import io
from lxml import objectify, etree
from lxml.etree import XMLSyntaxError
from PIL import Image
from pikepdf import Pdf, AttachedFileSpec, Name, String, Array, Dictionary, OutlineItem, Stream
import pikepdf
from pikepdf.models.metadata import (XMP_NS_DC, XMP_NS_PDF, XMP_NS_XMP, DateConverter, PdfMetadata, decode_pdf_date)
import pytest
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.pdfgen import canvas
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.pdfdoc import PDFCatalog, PDFDictionary, PDFName, PDFArray, PDFString, PDFIndirectObject
# from reportlab.pdfbase.pdfmetrics import registerFontFamily
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.lib.units import mm, cm, inch
from reportlab.lib.enums import TA_JUSTIFY, TA_LEFT, TA_CENTER, TA_RIGHT
from reportlab.platypus import Paragraph, Table, KeepInFrame, SimpleDocTemplate
from reportlab.lib import colors
# from PyPDF2 import PdfFileReader, PdfFileWriter
import textwrap
import pyodbc
from icecream import ic
import logging
import os
import qrcode
from pydantic import BaseModel
app = FastAPI()
app.mount("/webpages", StaticFiles(directory="webpages"), name="webpages")
laboratory = 'Sample lab'
app_version = 'Sample Program
temp_xml_file_path = None
xml_list = None
logged_in_user = None
signature_image = None
signature_bytes = None
temp_xsd_file_path = None
xsd_list = None
previewfile = None
accreditation_logo = True
class CheckboxUpdate(BaseModel):
    checked: bool
def connect_to_database(username: str, password: str):
…
@app.post("/create_account")
async def create_account(username: str = Form(...),
…
@app.post("/login")
async def login(username: str = Form(...), password: str = Form(...)):
…
@app.post("/handle_xml_click/{item_id}")
async def handle_xml_click(item_id: int):
…
@app.post("/handle_xsd_click/{item_id}")
async def handle_xsd_click(item_id: int):
…
def read_xml_list():
    with open("xml_list.db", "r") as file:
…
def read_xsd_list():
…
@app.get("/xml_list")
async def get_xml_list():
…
@app.get("/xsd_list")
async def get_xml_list():
…
@app.get("/uploadxmlfile")
async def main():
…
@app.post("/uploadxmlfile")
async def upload_xml_file(response: Response, xml_file: UploadFile = File(...)):
…
@app.get("/uploadxsdfile")
async def main():
…
@app.post("/uploadxsdfile")
async def upload_xsd_file(xsd_file: UploadFile = File(...)):
…
@app.post("/generatefile")
async def generate_file():
    global temp_xml_file_path, logged_in_user
    item_id: int
    if logged_in_user is None:
        return RedirectResponse(url="/login")

    xml_list_content = read_xml_list()
    clicked_item = xml_list_content[0].strip()
    file_path = eval(clicked_item)["FilePath"]
    xml_obj = get_xml_object(file_path)
    validate_xsd = validate_xml_against_xsd(file_path, temp_xsd_file_path)
    if not validate_xsd:
        print("Invalid XML file!")
        return HTTPException(status_code=400, detail="XML doesn't adhere to XSD")
    elif validate_xsd is None:
        print("No XSD file selected. Please select an XSD file.")
    elif validate_xsd:
        print("XSD file:", temp_xsd_file_path)
        print("xml file:", file_path)
        print("Valid XML file!")
        pdf_output = generate_pdf(previewfile=False, checkbox=accreditation_logo)
        pdf_with_xml = attach_xml_to_pdf(pdf_output, file_path)
        file1 = xml_obj.xpath('//ns1:file1', namespaces={'ns1': 'https://…'})
        itemname = xml_obj.xpath('//ns1:itemname', namespaces={'ns1': 'https://…'})
        fname = f"{file1[0]} - {itemname[0]}.pdf"
        with NamedTemporaryFile(delete=False, suffix=".pdf") as temp_output_file:
            pdf_with_xml.save(temp_output_file, force_version="1.7", preserve_pdfa=True, fix_metadata_version=True)
        checks = check_pdfa(temp_output_file.name)
        return FileResponse(temp_output_file.name, filename=fname)
@app.post("/previewfile")
async def generate_file():
…
@app.post("/accredited")
async def checkbox(update: CheckboxUpdate):
…
def validate_xml_against_xsd(file_path, temp_xsd_file_path):
…
def attach_xml_to_pdf(pdf_file, file_path):
    global temp_xml_file_path, laboratory
    xml_list_content = read_xml_list()
    clicked_item = xml_list_content[0].strip()
    file_path = eval(clicked_item)["FilePath"]
    xml_obj = get_xml_object(file_path)

    AnalystName = xml_obj.xpath('//ns1:AnalystName', namespaces={'ns1': 'https://…'})
    ApproverName = xml_obj.xpath('//ns1:ApproverName', namespaces={'ns1': 'https://…'})
    file1 = xml_obj.xpath('//ns1:file1', namespaces={'ns1': 'https://…'})
    itemname = xml_obj.xpath('//ns1:itemname', namespaces={'ns1': 'https://…'})
    refitem = [item.text for item in xml_obj.xpath('//ns1:refitem/ns1:Item', namespaces={'ns1': 'https://…'})]
    pdf = Pdf.new()
    filespec = AttachedFileSpec.from_filepath(pdf, file_path)
    attachment_name = '/' + file_path
    pdf.attachments[attachment_name] = filespec
    pdf_input =  Pdf.open(pdf_file)
    for page in pdf_input.pages:
        pdf.pages.append(page)
    with pdf_input.open_metadata(
        set_pikepdf_as_editor=False, update_docinfo=False
    ) as xmp_view:
        assert not xmp_view
    with pdf_input.open_metadata(
        set_pikepdf_as_editor=False, update_docinfo=False) as xmp:
        assert not xmp
    del xmp
    with pdf_input.open_metadata(
        set_pikepdf_as_editor=False, update_docinfo=False) as xmp:
            xmp['dc:title'] = 'Document'
            xmp['pdfaSchema:Keywords'] = ['Document', file1[0].text, itemname[0].text]
            xmp['dc:creator'] = [AnalystName[0].text, ApproverName[0].text]
            xmp['pdf:Producer'] = [laboratory]
            xmp['xmp:CreatorTool'] = [app_version]
            xmp['dc:STDUsed'] = [refitem]
            xmp['dc:publisher'] = [app_version]
            xmp['pdf:PDFVersion'] = '1.7'
            xmp['pdfaExtension'] = 'http://www.aiim.org/pdfa/ns/extension/'
            xmp['pdfaField'] = 'http://www.aiim.org/pdfa/ns/field#'
            xmp['pdfaProperty'] = 'http://www.aiim.org/pdfa/ns/property#'
            xmp['pdfaSchema'] = 'http://www.aiim.org/pdfa/ns/schema#'
            xmp['pdfaType'] = 'http://www.aiim.org/pdfa/ns/type#'
            assert '<pdf:Producer xmlns:pdf="http://ns.adobe.com/pdf/1.7/">'
    xmpstr1 = pdf_input.Root.Metadata.read_bytes()
    xpacket_begin = b'<?xpacket begin='
    xpacket_end = b'<?xpacket end='
    assert xmpstr1.startswith(xpacket_begin)

    with pdf_input.open_metadata() as xmp:
        xmp['dc:creator'] = ['Foo']

    xmpstr2 = pdf_input.Root.Metadata.read_bytes()
    assert xmpstr2.startswith(xpacket_begin)

    def only_one_substring(s, subs):
        return s.find(subs) == s.rfind(subs)

    assert only_one_substring(xmpstr2, xpacket_begin)
    assert only_one_substring(xmpstr2, xpacket_end)
    print(pdf_input.Root.Metadata.read_bytes())
def libxmp_meta():
    try:
        libxmp = pytest.importorskip('libxmp')
    except Exception as e:  # pylint: disable=broad-except
        # libxmp raises a nonstandard exception sometimes so the standard
        # pytest.importorskip doesn't trap it *irritated sigh*.
        pytest.skip(str(e))
    return libxmp.XMPMeta
def include_sections(pdf_with_xml):
…
def check_pdfa(temp_output_file):
…
def generate_pdf(previewfile, checkbox):
    pdf_output = BytesIO()
    c = canvas.Canvas(pdf_output, pagesize=A4)
    xml_list_content = read_xml_list()
    clicked_item = xml_list_content[0].strip()
    file_path = eval(clicked_item)["FilePath"]
    xml_obj = get_xml_object(file_path)
    width, height = A4
    styles = getSampleStyleSheet()
    
    stylesub = ParagraphStyle('stylesub',
                            fontName="Calibrib",
                            alignment=TA_LEFT,
                            fontSize=11,
                            parent=styles['Normal'])
…
    c.save()
    pdf_output.seek(0)

    pdf_output.getvalue()
    return pdf_output
```
below is the resulting metada based on the print function (print(pdf_input.Root.Metadata.read_bytes())):
```
b'<?xpacket begin="\xef\xbb\xbf" id="W5M0MpCehiHzreSzNTczkc9d"?>\n
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="pikepdf">\n
	<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">\n
		<rdf:Description rdf:about="">
			<dc:title xmlns:dc="http://purl.org/dc/elements/1.1/">
				<rdf:Alt>
					<rdf:li xml:lang="x-default">Calibration Certificate</rdf:li>
				</rdf:Alt>
			</dc:title>
		</rdf:Description>
	<rdf:Description rdf:about="">
		<pdfaSchema:Keywords xmlns:pdfaSchema="http://www.aiim.org/pdfa/ns/schema#">
			<rdf:Seq>
				<rdf:li>Document</rdf:li>
				<rdf:li>2023ID</rdf:li>
				<rdf:li>Item1</rdf:li>
			</rdf:Seq>
		</pdfaSchema:Keywords>
	</rdf:Description>
	<rdf:Description rdf:about="">
		<dc:creator xmlns:dc="http://purl.org/dc/elements/1.1/">
			<rdf:Seq>
				<rdf:li>Foo</rdf:li>
			</rdf:Seq>
		</dc:creator>
	</rdf:Description>
	<rdf:Description rdf:about="">
		<pdf:Producer xmlns:pdf="http://ns.adobe.com/pdf/1.3/">pikepdf 8.8.0</pdf:Producer>
	</rdf:Description>
	<rdf:Description rdf:about="">
		<xmp:CreatorTool xmlns:xmp="http://ns.adobe.com/xap/1.0/">
			<rdf:Seq>
				<rdf:li>Sample Program</rdf:li>
			</rdf:Seq>
		</xmp:CreatorTool>
	</rdf:Description>
	<rdf:Description rdf:about="">
		<dc:refitem xmlns:dc="http://purl.org/dc/elements/1.1/">
			<rdf:Seq>
				<rdf:li>list1; list2</rdf:li>
			</rdf:Seq>
		</dc:refitem>
	</rdf:Description>
	<rdf:Description rdf:about="">
		<dc:publisher xmlns:dc="http://purl.org/dc/elements/1.1/">
			<rdf:Seq>
				<rdf:li>Sample Program</rdf:li>
			</rdf:Seq>
		</dc:publisher>
	</rdf:Description>
	<rdf:Description rdf:about="" x:pdfaExtension="http://www.aiim.org/pdfa/ns/extension/"/>
	<rdf:Description rdf:about="" x:pdfaField="http://www.aiim.org/pdfa/ns/field#"/>
	<rdf:Description rdf:about="" x:pdfaProperty="http://www.aiim.org/pdfa/ns/property#"/>
	<rdf:Description rdf:about="" x:pdfaSchema="http://www.aiim.org/pdfa/ns/schema#"/>
	<rdf:Description rdf:about="" x:pdfaType="http://www.aiim.org/pdfa/ns/type#"/>
	<rdf:Description xmlns:xmp="http://ns.adobe.com/xap/1.0/" rdf:about="" xmp:MetadataDate="2023-12-12T10:28:30.782196+00:00"/>
</rdf:RDF>\n</x:xmpmeta>\n\n<?xpacket end="w"?>\n'
```

I was able to generate a pdf but it does not conform to PDF/A standard as per online validation tools such as [pdf-online.com](https://www.pdf-online.com/osa/validate.aspx) and [avepdf.com](https://avepdf.com/pdfa-validation). this is the report of avepdf:
```
<?xml version="1.0" encoding="UTF-8"?>
<ValidationReport>
    <VersionInformation ID="GdPicture.NET.14" Version="14.2.53" />
    <ValidationProfile Conformance="PDF/A" Part="1" Level="A" />
    <FileInfo FileName="2023ID - item1.pdf" FileSize="288634 bytes" />
    <ValidationResult IsCompliant="False" Statement="PDF file is not compliant with validation profile requirements." />
    <Details>
        <FailedChecks Count="23">
            <Check ID="MissingXMPMetadata" OccurenceCount="1">
                <Occurence Context="Document" Statement="Document XMP metadata is missing." ObjReference="None" />
            </Check>
            <Check ID="UnsupportedPDFA1EmbeddedFile" OccurenceCount="1">
                <Occurence Context="Document" Statement="Name dictionary contains prohibited EmbeddedFiles entry." ObjReference="None" />
            </Check>
            <Check ID="MissingMarkInfoDictionary" OccurenceCount="1">
                <Occurence Context="Document" Statement="MarkInfo dictionary is missing." ObjReference="None" />
            </Check>
            <Check ID="MissingStructTreeRootDictionary" OccurenceCount="1">
                <Occurence Context="Document" Statement="StructTreeRoot dictionary not found." ObjReference="None" />
            </Check>
            <Check ID="UnsupportedDeviceColorSpace" OccurenceCount="13">
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="None" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="14 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="11 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="13 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="5 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="15 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="12 0 obj" />
                <Occurence Context="Page" PageNumber="2" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="14 0 obj" />
                <Occurence Context="Page" PageNumber="2" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="11 0 obj" />
                <Occurence Context="Page" PageNumber="2" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="6 0 obj" />
                <Occurence Context="Page" PageNumber="3" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="14 0 obj" />
                <Occurence Context="Page" PageNumber="3" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="11 0 obj" />
                <Occurence Context="Page" PageNumber="3" Statement="Device dependent color space used without matching PDF/A OutputIntent." ObjReference="7 0 obj" />
            </Check>
            <Check ID="ProhibitedXObjectSMaskEntry" OccurenceCount="6">
                <Occurence Context="Page" PageNumber="1" Statement="The XObject contains SMask dictionary entry." ObjReference="14 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="The XObject contains SMask dictionary entry." ObjReference="11 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="The XObject contains SMask dictionary entry." ObjReference="15 0 obj" />
                <Occurence Context="Page" PageNumber="1" Statement="The XObject contains SMask dictionary entry." ObjReference="12 0 obj" />
                <Occurence Context="Page" PageNumber="2" Statement="The XObject contains SMask dictionary entry." ObjReference="11 0 obj" />
                <Occurence Context="Page" PageNumber="3" Statement="The XObject contains SMask dictionary entry." ObjReference="11 0 obj" />
            </Check>
        </FailedChecks>
    </Details>
</ValidationReport>
```

TL:DR:
1. I want to generate a PDF/A compliant document
2. I use 'open_metadata' to edit the xmp and attach the necessary information
3. the xmp reflects my changes as per print function, however, validation fails using online tools (MissingXMPMetadata, UnsupportedPDFA1EmbeddedFile, MissingMarkInfoDictionary, MissingStructTreeRootDictionary, UnsupportedDeviceColorSpace, ProhibitedXObjectSMaskEntry)
4. based on the report of validation, how can I properly include the xmp metadata and embed the colorspace? i have read the [sample codes](https://github.com/pikepdf/pikepdf/blob/main/tests) but still having difficulties and the embedding of colorspace into the PDF were not included in the tests file (or I just can't find it).

**Comment by jbarlow83** [comment]
Merely editing the XMP metadata does not create PDF/A compliance. You would have to inspect all components of the input PDF and confirm they comply with PDF/A, or convert them if they do not. Consider using Apache pdfbox or Ghostscript to perform PDF/A conversion. pikepdf could be used to create a PDF/A conversion utility, could be used to create all of the required data structures in the correct format, but it does not have built-in high level functions to perform PDF/A conversion.
